# # 소프트웨어_테스팅_7

생성 일시: 2023년 1월 25일 오후 12:22

## Spring Boot Test

스프링 테스트 모듈

- Junit
- Spring Test
- AssertJ → Assertion library
- Hamcrest → A library of Matcher objects
- Mockito → Java Mocking library
- JSONassert  → assertion library for JSon
- JSONPath

위 라이브러리 포함 하는 spring-boot-starter-test

### Unit Testing in Spring

**Mock Objects**

스프링으로 유닛테스트를 진행할 때는 특정 단일 유닛에서 다른 유닛을 참고하거나, 다른 계층(Controller, Repository 등)의 로직 등이 필요한 경우가 많다. 이 때, 실제로 구현되지 않은 다른 계층의 객체를 생성해 사용할 수 있도록 지원한다. 이를 통해 고립된 환경에서의 Unit Test를 실시 할 수 있다.

예를들어) 로그인 기능을 테스트한다고 가정해보자, 테스터는 다음과 같은 사항들을 테스트할 수 있다.

1. Invalid Id
2. Invalid Pw
3. Valid Id & Invalid Pw
4. Invalid Id & Valid Pw
5. **Already Exist Id**

위 사항 중에서 1,2,3,4 번 경우에 대해서는 다른 어떤 계층에 의존적이지 않고 독립적으로 테스트를 수행할 수 있을 것이다. 하지만, 5번의 경우 특정 아이디에 대해 이미 존재하는 지에 대한 여부를 확인할 때, User 정보들을 담고있는 Repository가 필요할 것이다. 이때 Mock객체를 활용할 수 있다.

Mock 객체(Member Repository)를 생성하여 아이디를 검사하는 로직을 수행하게 할 수 있다.

실제 객체를 가져오는 것에 비해 의존성을 많이 줄여 가벼운 테스트를 진행할 수 있다.

---

<aside>
🌱 **Mockito**

![Untitled](#%20%E1%84%89%E1%85%A9%E1%84%91%E1%85%B3%E1%84%90%E1%85%B3%E1%84%8B%E1%85%B0%E1%84%8B%E1%85%A5_%E1%84%90%E1%85%A6%E1%84%89%E1%85%B3%E1%84%90%E1%85%B5%E1%86%BC_7%206eebc842c6434c939dd7b0a489ae8bf2/Untitled.png)

Mockito 의존성 설정
Gradle 

dependency에 `testImplementation 'org.mockito:mockito-core:{version}'` 추가

Maven pom파일에

`<dependency>
<groupId>org.mockito</groupId>
<artifactId>mockito-core</artifactId>
<version>5.0.0</version>
<scope>test</scope>
</dependency>` 추가

Mock(ClassToMock.class)으로 Mock객체 생성 가능

Mock 객체의 내부코드는 모두 동작하지 않고 리턴 값은 0, false, null 등으로 대체된다.

When을 사용하면 특정 조건에서 특정 값을 리턴하도록 할 수 있다.

```java
class Test{
MockRepository = Mock(MemberRepository.class)
String id = "userId";

@Test
public void WhenUserIdAlreadExist_ThrowsEx()
{
	when(MockReposotory.FindById(id))
   .thenThrow(new AlreadExistsException());
}
}
```

Mock 객체를 활용해서 MemberRepository의 행동을 Mocking할 수 있다. 위의 예에서 MemberRepository의 FindById 메소드는 이미 존재하는 User의 id를 확인하여 이미 존재하는 유저를 알리기 위한 AlreadExistsException을 발생시키는 행동을 예상할 수 있고 실제로 MemberRepo를 구현하지 않고도 Mock을 통해 테스트를 진행 할 수 있다.

</aside>

### Mock 객체를 이용한 테스트 예제

![Untitled](#%20%E1%84%89%E1%85%A9%E1%84%91%E1%85%B3%E1%84%90%E1%85%B3%E1%84%8B%E1%85%B0%E1%84%8B%E1%85%A5_%E1%84%90%E1%85%A6%E1%84%89%E1%85%B3%E1%84%90%E1%85%B5%E1%86%BC_7%206eebc842c6434c939dd7b0a489ae8bf2/Untitled%201.png)

@InjectMocks 어노테이션

→ @Mock이 붙은 목 객체를 @InjcetMocks 어노테이션이 붙은 객체에 주입하는 어노테이션

위에서 calculatorService 목 객체를 mathApplication에 주입하였다.

![Untitled](#%20%E1%84%89%E1%85%A9%E1%84%91%E1%85%B3%E1%84%90%E1%85%B3%E1%84%8B%E1%85%B0%E1%84%8B%E1%85%A5_%E1%84%90%E1%85%A6%E1%84%89%E1%85%B3%E1%84%90%E1%85%B5%E1%86%BC_7%206eebc842c6434c939dd7b0a489ae8bf2/Untitled%202.png)

![Untitled](#%20%E1%84%89%E1%85%A9%E1%84%91%E1%85%B3%E1%84%90%E1%85%B3%E1%84%8B%E1%85%B0%E1%84%8B%E1%85%A5_%E1%84%90%E1%85%A6%E1%84%89%E1%85%B3%E1%84%90%E1%85%B5%E1%86%BC_7%206eebc842c6434c939dd7b0a489ae8bf2/Untitled%203.png)

Mock으로 사용하는 객체인 CalculatorService 객체의 구체적인 메소드는 구현하지 않은 것을 확인할 수 있다.

MathApplication은 Calculatory의 서비스 로직을 받아 실행된다.

@Test에서는 when이라는 mockito 라이브러리를 이용하여 구현되어 있지 않던 목 객체의 메서드를 정의할 수 있다.

when(CaculatorService.add(10,10)).thenReturn(20) → 10과 10을 더하는 기능임을 나타낸다.

이렇게 하면 목 객체는 10과 10이라는 인풋을 받아 20을 반환하도록 설정된다.

이 목 객체를 주입받은 MathApplication은 Junit의 Assert 라이브러리를 통해 테스트를 검증할 수 있게된다.

하지만, 이렇게 검증하는 것은 10과 10이라는 인풋을 넣었을 때 20이라는 올바른 값을 반환하는 기능정도만 검증할 수있다.

즉, 실제로 MathApplication에서 CaculatorService객체의 CaculatorService.add()메서드를 수행했는지 검증할 수는 없다.

이것을 확인하기 위해 verify()메서드를 이용한다.

![Untitled](#%20%E1%84%89%E1%85%A9%E1%84%91%E1%85%B3%E1%84%90%E1%85%B3%E1%84%8B%E1%85%B0%E1%84%8B%E1%85%A5_%E1%84%90%E1%85%A6%E1%84%89%E1%85%B3%E1%84%90%E1%85%B5%E1%86%BC_7%206eebc842c6434c939dd7b0a489ae8bf2/Untitled%204.png)

![Untitled](#%20%E1%84%89%E1%85%A9%E1%84%91%E1%85%B3%E1%84%90%E1%85%B3%E1%84%8B%E1%85%B0%E1%84%8B%E1%85%A5_%E1%84%90%E1%85%A6%E1%84%89%E1%85%B3%E1%84%90%E1%85%B5%E1%86%BC_7%206eebc842c6434c939dd7b0a489ae8bf2/Untitled%205.png)

MathApplication의 add 수행을 단지 인풋을 받아 두개의 값을 더해 결과를 반환하는 것으로 수정하고 Test에서 mathApplication의 행동을 검증하는 Verify()메서드를 사용해서 검증했다.

assertEquals를 통해 mathApplication의 결과값이 20임을 증명이 됐지만 실제로는 10과 10을 받아 20을 반환했기 때문에 통과가 된다.

하지만, mathApplication에서 CaculatorService의 기능을 통해 결과를 제공해야하는 로직을 어기게 되는 것을 검증하지 못했다.

이때 verify()를 통해 mock객체인 calculatorService의 add함수가 호출되었는지를 검사하는 문장을 추가했다.

결과는 👇

![Untitled](#%20%E1%84%89%E1%85%A9%E1%84%91%E1%85%B3%E1%84%90%E1%85%B3%E1%84%8B%E1%85%B0%E1%84%8B%E1%85%A5_%E1%84%90%E1%85%A6%E1%84%89%E1%85%B3%E1%84%90%E1%85%B5%E1%86%BC_7%206eebc842c6434c939dd7b0a489ae8bf2/Untitled%206.png)

예상대로 실패했다. calculatorService의 add함수가 호출되기를 기대했지만 실제로 zero interactions 즉, 함수가 호출되지 않았기 때문이다. 이처럼 verify를 통해 올바른 로직대로 움직이고 있는지를 검증할 수 있었다.